Updates deployment docs (#1383)
wasp-lang/wasp: web/docs/advanced/deployment/manually.md

---
title: Deploying Manually
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import AddExternalAuthEnvVarsReminder from './_addExternalAuthEnvVarsReminder.md'
import BuildingTheWebClient from './_building-the-web-client.md'
import SlashWarning from './_slash-warning.md'

You can deploy a Wasp project by generating the code and then deploying generated code manually. When a Wasp app is generated and built, it consists of three parts:
- a Node.js server
- a static client
- and a PostgreSQL database.

Anywhere where you can deploy Node.js apps or static apps, you can deploy your Wasp app. For example, you can deploy your client on [Netlify](https://www.netlify.com/), the server on [Fly.io](https://fly.io/) and the database on [Neon](https://neon.tech/).

## Deploying a Wasp App

To deploy a Wasp application:
1. We need to generate deployable code
1. Deploy the API server (backend)
1. Deploy the web client (frontend)
1. Have a PostgreSQL database running

Let's go through each of these steps.

### 1. Generating Deployable Code

> Run `wasp build` to generate deployable code for the whole app in the `.wasp/build/` directory.

Reviewer: ```suggestion
Running the command `wasp build` generates deployable code for the whole app in the `.wasp/build/` directory.
```

```
wasp build
```

:::caution PostgreSQL in production

> You will not be able to build the app if you are using SQLite as a database (which is a default database), you will have to [switch to PostgreSQL](#). <!-- TODO: update the link -->

Reviewer: ```suggestion
You won't be able to build the app if you are using SQLite as a database (which is the default database).
You'll have to [switch to PostgreSQL](#) before deploying to production. <!-- TODO: update the link -->
```

:::

### 2. Deploying API Server

> In `.wasp/build`, a `Dockerfile` is describing an image for building the server.
> 
> To run the server in production, deploy this Docker image to some hosting provider, ensure that environment variables are correctly set, and that is it.

Reviewer: ```suggestion
There's a Dockerfile that defines an image for building the server in the `.wasp/build` directory.

To run the server in production, deploy this Docker image to a hosting provider and ensure it has access to correct environment variables (this varies depending on the provider).
All necessary environment variables are listed in the next section.
```

#### Environment Variables

> The server uses the following environment variables, so you need to ensure they are set on your hosting provider:
> 
> - `PORT`
> 
>    The port number at which it will listen for requests, for example: `3001`
> 
> - `DATABASE_URL`
> 
>    The URL of the Postgres database the app should use, for example:
>    `postgresql://mydbuser:mypass@localhost:5432/nameofmydb`
> 
> - `WASP_WEB_CLIENT_URL`
> 
>    The URL of where the frontend app is running, for example: `https://<app-name>.netlify.app`), which is necessary for CORS
> 
> - `JWT_SECRET`
> 
>    You need this if you are using Wasp's `auth` feature. Set it to a random string at least 32 characters long (use an [online generator](https://djecrety.ir/))

Reviewer: ```suggestion
Here are the environment variables your server requires to run:

- `PORT`

  The server's HTTP port number. This is where the server listens for requests (e.g., `3001`).

- `DATABASE_URL`

  The URL of the Postgres database you want your app to use (e.g., `postgresql://mydbuser:mypass@localhost:5432/nameofmydb`).

- `WASP_WEB_CLIENT_URL`

  The URL where you plan to deploy your frontend app is running (e.g., `https://<app-name>.netlify.app`).
  The server needs to know about it to properly configure Same-Origin Policy (CORS) headers.

- `JWT_SECRET`

  You only need this environment variable if you're using Wasp's `auth` features.
  Set it to a random string at least 32 characters long (you can use an [online generator](https://djecrety.ir/)).
```

<AddExternalAuthEnvVarsReminder />

### 3. Deploying the Web Client

<BuildingTheWebClient />

The command above will build the web client (using Vite) and put it in the `build/` directory in the `web-app` directory.

> Since it's just a static client, you can deploy it to any static hosting provider.

Reviewer: ```suggestion
Since the app's frontend is just a bunch of static files, you can deploy it to any static hosting provider.
```

Additionally, `fly` is a symlink for `flyctl` on most systems and they can be used interchangeably.
:::

Make sure you are logged in with `flyctl` CLI. You can check if you are logged in with `flyctl auth whoami`, and if you are not, you can log in with `flyctl auth login`.

### Set Up a Fly.io App

:::info 
You need to do this only once per Wasp app.
:::

Unless you already have a Fly.io app that you want to deploy to, let's create a new Fly.io app. 

After you have [built the app](#1-generating-deployable-code), position yourself in `.wasp/build/` directory:

```shell
cd .wasp/build
```

Next, run the launch command to set up a new app and create a `fly.toml` file:

```bash
flyctl launch --remote-only
```

This will ask a series of questions, including what region to deploy in and if you would like a database.

- Say **yes** to **Would you like to set up a Postgresql database now?**, and select **Development**, and Fly.io will set a `DATABASE_URL` for you.
- Say **no** to **Would you like to deploy now?**, as well as any additional questions.
  
  We still need to set a few environment variables.

> :::info If the database setup fails

Reviewer: ```suggestion
:::info What if the database setup fails?
```

> If your attempts to initiate a new app fail for whatever reason, then you can run `flyctl apps destroy <app-name>` before trying again.

Author: Yes 👍

<details>
  <summary>
    What does it look like when your DB is deployed correctly?
  </summary>
  <div>

>     <p>When your DB is deployed correctly, you will be able to view it in the <a href="https://fly.io/dashboard">Fly.io dashboard</a>:</p>

Reviewer: ```suggestion
    <p>When your DB is deployed correctly, you'll see it in the <a href="https://fly.io/dashboard">Fly.io dashboard</a>:</p>
```
